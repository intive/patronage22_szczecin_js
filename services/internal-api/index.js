import axios from 'axios'

const axiosInstance = axios.create({
  baseURL: `https://${process.env.NEXT_PUBLIC_VERCEL_URL}` || 'http://localhost:3000',
  responseType: 'json',
  withCredentials: true
})

export const getBoards = async (...args) => {
  return axiosInstance.get('/api/boards', { params: Object.assign({}, ...args) })
}
