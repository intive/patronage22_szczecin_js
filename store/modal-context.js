import { createContext } from 'react'

const ModalContext = createContext({})

export const ModalProvider = ModalContext.Provider

export default ModalContext
