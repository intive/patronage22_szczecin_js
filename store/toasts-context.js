import { createContext } from 'react'
import useToast from '../hooks/useToast'

const ToastsContext = createContext({
  showSuccessToast: () => {},
  showErrorToast: () => {}
})

export const ToastsContextProvider = (props) => {
  const { showToast, RenderToast, isOpenToast } = useToast()

  const showSuccessToastHandler = async (notification) => {
    showToast('success', notification)
  }

  const showErrorToastHandler = async (notification) => {
    showToast('error', notification)
  }

  const context = {
    showSuccessToast: showSuccessToastHandler,
    showErrorToast: showErrorToastHandler
  }

  return (
    <ToastsContext.Provider value={context}>
      {props.children}
      {isOpenToast && <RenderToast />}
    </ToastsContext.Provider>
  )
}

export default ToastsContext
