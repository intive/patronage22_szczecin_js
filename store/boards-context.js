import { createContext, useState } from 'react'
import { getBoards } from '../services/internal-api'

const BoardsContext = createContext({
  boards: [],
  reload: () => {}
})

export const BoardsContextProvider = (props) => {
  const [updateBoards, setUpdateBoards] = useState(props.boards)

  const reloadHandler = async () => {
    setUpdateBoards((await getBoards()).data)
  }

  const context = {
    boards: updateBoards,
    reload: reloadHandler
  }

  return (
    <BoardsContext.Provider value={context}>
      {props.children}
    </BoardsContext.Provider>
  )
}

export default BoardsContext
