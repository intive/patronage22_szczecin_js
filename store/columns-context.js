import { createContext, useState } from 'react'
import { getBoard } from '../services/internal-api'

const ColumnsContext = createContext({
  columns: [],
  reload: () => {}
})

export const ColumnsContextProvider = (props) => {
  const [updateColumns, setUpdateColumns] = useState(props.columns || [])

  const reloadHandler = async (id) => {
    try {
      const { data } = await getBoard(id)
      setUpdateColumns(data.columns)
    } catch (err) {
      console.log(err)
    }
  }

  const context = {
    columns: updateColumns,
    reload: reloadHandler
  }

  return (
    <ColumnsContext.Provider value={context}>
      {props.children}
    </ColumnsContext.Provider>
  )
}

export default ColumnsContext
