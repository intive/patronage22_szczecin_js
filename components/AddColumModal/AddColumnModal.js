import Modal from '../Modal/Modal'
import { useTranslation } from 'next-i18next'
import { StyledInput } from './style'
import { useState, useContext, useEffect } from 'react'
import { addColumn } from '../../services/internal-api'
import ColumnsContext from '../../store/columns-context'
import ToastsContext from '../../store/toasts-context'
import { ModalProvider } from '../../store/modal-context.js'

export default function AddColumnModal ({ boardId, handleClose, isModalOpen }) {
  const { t } = useTranslation(['common', 'modals'])
  const columnsCtx = useContext(ColumnsContext)
  const toastsCtx = useContext(ToastsContext)
  const [columnName, setColumnName] = useState('')

  const toastMessage = {
    success: t('toastMessage.column.successToast'),
    error: t('toastMessage.errorToast')
  }
  const addColumnHandler = async (id, name) => {
    try {
      await addColumn(id, name)
      toastsCtx.showSuccessToast(toastMessage.success)
      columnsCtx.reload(boardId)
    } catch (error) {
      toastsCtx.showErrorToast(toastMessage.error)
    }
  }

  const handleOnChange = (event) => {
    setColumnName(event.target.value)
  }

  const handleSaveClick = () => {
    addColumnHandler(boardId, columnName)
  }

  useEffect(() => {
    if (!isModalOpen) {
      setColumnName('')
    }
  }, [isModalOpen])

  const modalOptions = {
    isModalOpen,
    handleClose,
    handleSaveClick,
    title: t('addColumnModal.title', { ns: 'modals' }),
    subtitle: t('addColumnModal.subtitle', { ns: 'modals' }),
    disabled: columnName.length < 5
  }

  return (
    <ModalProvider value={modalOptions}>
      <Modal>
        <StyledInput value={columnName} onChange={handleOnChange} maxLength='30' />
      </Modal>
    </ModalProvider>
  )
}
