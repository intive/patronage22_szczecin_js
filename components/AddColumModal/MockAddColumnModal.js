import Modal from '../Modal/Modal'
import { useTranslation } from 'next-i18next'
import { StyledInput } from './style'
import { useContext } from 'react'
import ToastsContext from '../../store/toasts-context'
import { ModalProvider } from '../../store/modal-context.js'

export default function MockAddColumnModal ({ id, isModalOpen, handleClose }) {
  const { t } = useTranslation(['common', 'modals'])
  const toastsCtx = useContext(ToastsContext)

  const toastMessage = {
    success: t('toastMessage.column.successToast'),
    error: t('toastMessage.errorToast')
  }
  const addColumnHandler = async (id) => {
    const data = { status: id }

    if (data.status === 200) {
      toastsCtx.showSuccessToast(toastMessage.success)
    } else {
      toastsCtx.showErrorToast(toastMessage.error)
    }
  }

  const handleSaveClick = () => {
    addColumnHandler(id)
  }

  const modalOptions = {
    isModalOpen,
    handleClose,
    handleSaveClick,
    title: t('addColumnModal.title', { ns: 'modals' }),
    subtitle: t('addColumnModal.subtitle', { ns: 'modals' })
  }

  return (
    <ModalProvider value={modalOptions}>
      <Modal>
        <StyledInput maxLength='30' />
      </Modal>
    </ModalProvider>
  )
}
