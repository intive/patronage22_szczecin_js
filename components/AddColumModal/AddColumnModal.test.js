/* globals describe, it, expect, jest */

import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import AddColumnModal from './AddColumnModal'
import * as InternalApi from '../../services/internal-api'
import { ToastsContextProvider } from '../../store/toasts-context'
import MockAddColumnModal from './MockAddColumnModal'

describe('AddColumnModal', () => {
  const mockAddColumn = jest.spyOn(InternalApi, 'addColumn')

  it('should render AddColumnModal correctly', () => {
    const { baseElement } = render(<AddColumnModal isModalOpen='true' />)

    expect(baseElement).toMatchSnapshot()
  })

  it('should call addColumn option when save button is clicked', () => {
    const mockOnClick = jest.fn()
    render(<AddColumnModal boardId='mockedId' isModalOpen handleClose={mockOnClick} />)
    expect(screen.getByRole('button', { name: /buttons.save/i })).toBeInTheDocument()
    userEvent.type(screen.getByRole('textbox'), 'Column1')
    expect(screen.getByRole('textbox')).toHaveValue('Column1')
    expect(screen.queryByRole('button', { name: /buttons.save/i })).not.toBeDisabled()
    userEvent.click(screen.getByRole('button', { name: /buttons.save/i }))
    expect(mockAddColumn).toHaveBeenCalledTimes(1)
    expect(mockAddColumn).toHaveBeenCalledWith('mockedId', 'Column1')
  })

  it('should display a success toast after a successful add column operation', async () => {
    const mockOnClick = jest.fn()
    const mockStatus = mockAddColumn.mockResolvedValue({ status: 200 })
    const data = await mockStatus()

    render(<ToastsContextProvider><MockAddColumnModal isModalOpen handleClose={mockOnClick} id={data.status} /></ToastsContextProvider>)

    userEvent.type(screen.getByRole('textbox'), 'Column1')
    userEvent.click(screen.getByRole('button', { name: /buttons.save/i }))

    expect(screen.getByText(/toastMessage.column.successToast/)).toBeInTheDocument()
  })

  it('should display an error toast after a failed add column operation', async () => {
    const mockOnClick = jest.fn()
    const mockStatus = mockAddColumn.mockResolvedValue({ status: 404 })
    const data = await mockStatus()

    render(<ToastsContextProvider><MockAddColumnModal isModalOpen handleClose={mockOnClick} id={data.status} /></ToastsContextProvider>)

    userEvent.type(screen.getByRole('textbox'), 'Column1')
    userEvent.click(screen.getByRole('button', { name: /buttons.save/i }))

    expect(screen.getByText(/toastMessage.errorToast/)).toBeInTheDocument()
  })
})
