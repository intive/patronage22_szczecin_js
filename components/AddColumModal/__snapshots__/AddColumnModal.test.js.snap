// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddColumnModal should render AddColumnModal correctly 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    id="modal-portal"
  >
    <div
      class="sc-dkPtRN hBZupT"
      data-testid="backdrop"
    />
    <div
      aria-hidden="false"
      class="sc-hKwDye hLjaqj"
      data-testid="modal"
      role="dialog"
    >
      <div
        class="sc-eCImPb jdRPcx"
      >
        <p
          class="sc-gKclnd eVXqrc"
          id="modalTitle"
        >
          addColumnModal.title
        </p>
        <p
          class="sc-iCfMLu hKohOx"
          id="modalSubtitle"
        >
          addColumnModal.subtitle
        </p>
      </div>
      <div
        class="sc-furwcr coHOBm"
      >
        <input
          class="sc-iqseJM hgSgod sc-egiyK cJcWzW"
          maxlength="30"
          value=""
        />
        <span
          class="sc-crHmcD dpZhDT"
          data-testid="character-counter"
        >
          0
          /
          30
        </span>
      </div>
      <div
        class="sc-pVTFL ldmNEs"
      >
        <button
          class="sc-bdvvtL jVCSJw sc-jrQzAO bsfgRs"
        >
          buttons.cancel
        </button>
        <button
          class="sc-bdvvtL dRBcfm sc-kDTinF gcMRAr"
          disabled=""
        >
          buttons.save
        </button>
      </div>
    </div>
  </div>
</body>
`;
